olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -769,6 +894,77 @@ client::CancellationToken AuthenticationClient::Impl::SignOut(
   });
 }
 
+client::CancellationToken AuthenticationClient::Impl::IntrospectApp(
+    std::string access_token, IntrospectAppCallback callback) {
+  auto introspect_app_task = [=](client::CancellationContext context)
+      -> client::ApiResponse<IntrospectAppResult, client::ApiError> {
+    if (!settings_.network_request_handler) {
+      client::ApiError result({static_cast<int>(http::ErrorCode::IO_ERROR),
+                               "Cannot introspect app while offline"});
+      return std::move(result);
+    }
+    std::string url = settings_.token_endpoint_url + kIntrospectAppEndpoint;
+    auto network_settings = http::NetworkSettings().WithProxySettings(
+        settings_.network_proxy_settings.value_or(
+            http::NetworkProxySettings{}));
+    auto request =
+        http::NetworkRequest(url)
+            .WithVerb(http::NetworkRequest::HttpVerb::GET)
+            .WithHeader(http::kAuthorizationHeader,
+                        generateBearerHeader(access_token))
+            .WithHeader(http::kUserAgentHeader, http::kOlpSdkUserAgent)
+            .WithSettings(std::move(network_settings));
+
+    auto payload = std::make_shared<std::stringstream>();
+    auto send_outcome = settings_.network_request_handler->Send(
+        request, payload,
+        [callback, payload](const http::NetworkResponse& network_response) {
+          auto response_status = network_response.GetStatus();
+          auto error_msg = network_response.GetError();
+
+          if (response_status != 200) {
+            // Network error response not available
+            AuthenticationError result(response_status, error_msg);
+            callback(std::move(result));
+            return;
+          }
+
+          auto document = std::make_shared<rapidjson::Document>();
+          rapidjson::IStreamWrapper stream(*payload);
+          document->ParseStream(stream);
+
+          IntrospectAppResult result =
+              AuthenticationClient::Impl::parseIntrospectAppResult(document);
+          callback(std::move(result));
+        });
+
+    if (!send_outcome.IsSuccessful()) {
+      std::string error_message =
+          ErrorCodeToString(send_outcome.GetErrorCode());
+      client::ApiError result({static_cast<int>(send_outcome.GetErrorCode()),
+                               std::move(error_message)});
+      return std::move(result);
+    }
+    return IntrospectAppResult();
```

Comment:
Why dont you return the real result and call the callback from the below wrap_callback? Why do you bind the callback to two lambdas?

Reply:
Inside of wrap_callback we can't distinguish valid result from network->Send() response and empty result returned from introspect_app_task completion (`return IntrospectAppResult();`).

Reply:
But why are you returning an empty `IntrospectAppResult()` back? Is it because your wrap_callback has the assumption that only errors are sent to it. So if you also send the parsed result to wrap_callback, do you still need to return the default constructed `IntrospectAppResult`?

Reply:
network->Send() is async, parsed result is not ready when finishing introspect_app_task.
So I can only return some error or an empty result.
Well, I can use some special error and handle this case in wrap_callback.

Reply:
Maybe it is then better to use OlpClient::CallApi(), as this is sync. If it is not possible then ignore my comment, although it looks very confusing.

Reply:
Done, should be better now.